src/Upload/StreamableUploader.php

```diff
@@ -0,0 +1,165 @@
+<?php
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may ob`tain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Upload;
+
+use GuzzleHttp\Psr7\BufferStream;
+use GuzzleHttp\Psr7\Request;
+
+/**
+ * Uploader that is a special case of the ResumableUploader where we can write
+ * the file contents in a streaming manner.
+ */
+class StreamableUploader extends ResumableUploader
```

Comment:
Yes, originally I had tried implementing it inside as a special case. However, it really felt like the class was doing 2 distinct things. Yes, they both use the resumable api, but the mechanism for handling the input data was different.

I tried using an `AppendStream` as the `$data`, but if you keep writing new data, your list of streams grows. Ideally, once we've sent a data from the stream, we can free up the memory (which is why I opted for using a `BufferStream` then replacing it after we have written a chunk. In the normal `ResumableUploader` case, this isn't a thing to do.

I could also write an improved appending stream that could marginally improve the finding the chunks to send and/or freeing memory (at the cost of the ability to seek).